samples/api/swapchain_recreation/swapchain_recreation.cpp

```diff
@@ -0,0 +1,950 @@
+	LOGI("Fence pool size at destruction: {}", fence_pool.size());
+
+	for (VkSemaphore semaphore : semaphore_pool)
+	{
+		vkDestroySemaphore(get_device_handle(), semaphore, nullptr);
+	}
+
+	for (VkFence fence : fence_pool)
+	{
+		vkDestroyFence(get_device_handle(), fence, nullptr);
+	}
+
+	cleanup_swapchain_objects(swapchain_objects);
+	if (swapchain != VK_NULL_HANDLE)
+	{
+		vkDestroySwapchainKHR(get_device_handle(), swapchain, nullptr);
+	}
+
+	if (render_pass != VK_NULL_HANDLE)
+	{
+		vkDestroyRenderPass(get_device_handle(), render_pass, nullptr);
+	}
+}
+
+bool SwapchainRecreation::prepare(vkb::Platform &platform)
+{
+	return VulkanSample::prepare(platform);
+}
+
+void SwapchainRecreation::create_render_context(vkb::Platform &platform)
+{
+	get_queue();
+	query_surface_format();
+	create_render_pass();
+	init_swapchain();
+}
+
+void SwapchainRecreation::prepare_render_context()
+{
+	// Nothing to do
+}
+
+void SwapchainRecreation::update(float delta_time)
+{
+	fps_timer += delta_time;
+	if (fps_timer > 1.0f)
+	{
+		LOGI("FPS: {}", static_cast<float>(frame_number - fps_last_logged_frame_number) / fps_timer);
+		fps_timer -= 1.0f;
+		fps_last_logged_frame_number = frame_number;
+	}
+
+	++frame_number;
+
+	setup_frame();
+
+	if (desired_present_mode != current_present_mode)
+	{
+		recreate_swapchain();
+	}
+
+	uint32_t index;
+	auto     res = acquire_next_image(&index);
+
+	// Handle outdated error in acquire.
+	if (res == VK_SUBOPTIMAL_KHR || res == VK_ERROR_OUT_OF_DATE_KHR)
+	{
+		recreate_swapchain();
+		res = acquire_next_image(&index);
+	}
+	if (res != VK_SUBOPTIMAL_KHR)
+	{
+		VK_CHECK(res);
+	}
+
+	render(index);
+	res = present_image(index);
+
+	// Handle Outdated error in present.
+	if (res == VK_SUBOPTIMAL_KHR || res == VK_ERROR_OUT_OF_DATE_KHR)
+	{
+		recreate_swapchain();
+	}
+	else
+	{
+		VK_CHECK(res);
+	}
+}
+
+bool SwapchainRecreation::resize(const uint32_t, const uint32_t)
+{
+	if (get_device_handle() == VK_NULL_HANDLE)
+	{
+		return false;
+	}
+
+	return recreate_swapchain();
+}
+
+void SwapchainRecreation::input_event(const vkb::InputEvent &input_event)
+{
+	if (input_event.get_source() != vkb::EventSource::Keyboard)
+	{
+		return;
+	}
+
+	const auto &key_button = static_cast<const vkb::KeyInputEvent &>(input_event);
+	if (key_button.get_action() != vkb::KeyAction::Up)
+	{
+		return;
+	}
+
+	switch (key_button.get_code())
+	{
+		case vkb::KeyCode::V:
+			// Note: events are being double-sent, avoid double logging with this check
+			// as a workaround.
+			if (current_present_mode != VK_PRESENT_MODE_FIFO_KHR)
+			{
+				LOGI("Enabling V-Sync");
+				desired_present_mode = VK_PRESENT_MODE_FIFO_KHR;
+			}
+			break;
+		case vkb::KeyCode::N:
+			if (current_present_mode == VK_PRESENT_MODE_FIFO_KHR)
+			{
+				LOGI("Disabling V-Sync");
+				desired_present_mode = VK_PRESENT_MODE_MAILBOX_KHR;
+			}
+			break;
+		default:
+			break;
+	}
+
+	query_present_modes();
```

Comment:
You'd be surprised (as I was), but apparently dragging a window through monitors or something like that can affect supported present modes.

It came up during VK_EXT_swapchain_maintenance1 that there's (an existing) race condition about that (swapchain created successfully with one present mode, then fails present right after because present modes changed)